### PR TITLE
Add template: Barracuda WAF with IIS Backend Servers

### PR DESCRIPTION
**Pull request by amalik99**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

*
*
*

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@amalik99__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

**Comment by amalik99** [comment]
Hi, Adding _ci_skip file for following reasons

1.  Template use marketplace item "Barracuda Web Application Firewall" , for which programmatic deployment must be enabled in the subscription. Raw error  " **failed validation to purchase resources. Error message: 'Legal terms have not been accepted for this item on this subscription. To accept legal terms, please go to the Azure portal (http://go.microsoft.com/fwlink/?LinkId=534873) and configure programmatic deployment for the Marketplace item or create it there for the first time'\n",**

2. On line 562, Actual Code will be **"ModulesUrl": "[concat(parameters('_artifactsLocation'),'scripts/',variables('dscfilename'),parameters('_artifactsLocationSasToken'))]",** . Travis CI isnt able to modify this uri to use user branch of github on runtime , have used direct URI for travis validation temprorarily. Direct URI **"ModulesUrl": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/barracuda-waf-iis/scripts/webserverconfig.zip",**

3._artifactsLocationSasToken value is blank which is as expected, Travis CI validation highlighting it as there's no value for this parameter. This is expected.

Please let me know incase of questions.

File: barracuda-waf-iis/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+The Barracuda Web Application Firewall inspects inbound web traffic and blocks SQL injections, Cross-Site Scripting, malware uploads & application DDoS and other attacks targeted at your web applications. It also inspects the responses from the back-end web servers for Data Loss Prevention (DLP). The integrated access control engine enables administrators to create granular access control policies for Authentication, Authorization & Accounting (AAA), which gives organizations strong authentication and user control. The onboard L4/L7 Load Balancing capabilities enable organizations to quickly add back-end servers to scale deployments. Application acceleration capabilities including SSL Offloading, caching, compression, and connection pooling, ensure faster application delivery of web application content. 
+
+Once deployment finishes, you will able to directly access fully configured Barracuda GUI and start publishing your web applications.
+
+##Deployment Solution Architecture 
+
+This template will deploy: 
+
+- Two storage account
```
Comment (bmoore-msft) [inline barracuda-waf-iis/README.md]: typo accounts s/b plural

File: barracuda-waf-iis/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,653 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "adminUsername": {
+      "type": "string",
+      "defaultValue": "bwafadmin",
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: remove default value - increases surface area

File: barracuda-waf-iis/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+      }
+    },
+    "webVmCount": {
+      "type": "int",
+      "defaultValue": 2,
+      "metadata": {
+        "description": "Number of backend Web servers to be deployed"
+      }
+    },
+    "webVmSize": {
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move to variable and use a size available in all clouds (Dv2 series)

File: barracuda-waf-iis/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+        "Standard_A1",
+        "Standard_A2",
+        "Standard_A3",
+        "Standard_A4"
+      ],
+      "metadata": {
+        "description": "Web VM size."
+      }
+    },
+    "bwafVmSize": {
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move to variable and use a size available in all clouds

File: barracuda-waf-iis/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,653 @@
...
+        "Standard_A4",
+        "Standard_A5",
+        "Standard_A6"
+
+      ],
+      "metadata": {
+        "description": "Barracuda WAF VM size."
+      }
+    },
+    "storageAccountType": {
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move to variable

File: barracuda-waf-iis/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+    "webNSGName": "web-NSG",
+    "bwafNSGName": "bwaf-NSG",
+    "bwafSubnetName": "bwaf-subnet",
+    "webSubnetName": "web-subnet",
+    "lbIPAddressName": "lb-pip",
+    "lbDnsLabel": "[concat('lbpip',uniqueString(resourceGroup().id))]",
+    "loadBalancerName": "web-rdp-lb",
+    "loadBalancerID": "[resourceId('Microsoft.Network/loadBalancers',variables('loadBalancerName'))]",
+    "loadBalancerIPID": "[resourceId('Microsoft.Network/publicIPAddresses',variables('lbIPAddressName'))]",
+    "frontEndIPConfigID": "[concat(variables('loadBalancerID'),'/frontendIPConfigurations/loadBalancerFrontEnd')]",
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: use resourceId()

File: barracuda-waf-iis/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+    "backEndIPConfigID": "[concat(variables('loadBalancerID'),'/ipConfigurations/ipconfig1')]",
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: use resourceId()

File: barracuda-waf-iis/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,653 @@
...
+    "vmStorageAccountContainerName": "vhds",
+    "OSDiskName": "OSDisk"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Storage/storageAccounts",
+      "name": "[variables('bwafStorageAccountName')]",
+      "apiVersion": "2016-01-01",
+      "location": "[resourceGroup().location]",
+      "comments": "Storage account  used to store the BWAF VM disks",
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: extra space

File: barracuda-waf-iis/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,653 @@
...
+      "sku": {
+        "name": "[parameters('storageAccountType')]"
+      },
+      "kind": "Storage",
+      "properties": {
+
+      },
+      "tags": {
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move tags under comments property

File: barracuda-waf-iis/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,653 @@
...
+      "location": "[resourceGroup().location]",
+      "comments": "Storage account used to store the Web VM disks",
+      "sku": {
+        "name": "[parameters('storageAccountType')]"
+      },
+      "kind": "Storage",
+      "properties": {
+
+      },
+      "tags": {
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move tags under comments

File: barracuda-waf-iis/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,653 @@
...
+      "location": "[resourceGroup().location]",
+      "comments": "Storage account used to store VM diagnostics data",
+      "sku": {
+        "name": "Standard_LRS"
+      },
+      "kind": "Storage",
+      "properties": {
+
+      },
+      "tags": {
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move tags under comments

File: barracuda-waf-iis/azuredeploy.json [hunk 12]
```diff
@@ -0,0 +1,653 @@
...
+              "sourceAddressPrefix": "Internet",
+              "destinationAddressPrefix": "10.0.0.0/24",
+              "access": "Allow",
+              "priority": 130,
+              "direction": "Inbound"
+            }
+          }
+        ]
+      },
+      "tags": {
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move tags under location property

(do the same on all the other resources too)

File: barracuda-waf-iis/azuredeploy.json [hunk 13]
```diff
@@ -0,0 +1,653 @@
...
+      "type": "Microsoft.Network/networkSecurityGroups",
+      "name": "[variables('bwafNSGName')]",
+      "apiVersion": "2016-03-30",
+      "location": "[resourceGroup().location]",
+      "properties": {
+        "securityRules": [
+          {
+            "name": "HTTP-allow",
+            "properties": {
+              "protocol": "TCP",
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: run this through schema validation (e.g. in VS Code) and fix what's reported

File: barracuda-waf-iis/azuredeploy.json [hunk 14]
```diff
@@ -0,0 +1,653 @@
...
+            "storageUri": "[reference(resourceId('Microsoft.Storage/storageAccounts/', variables('diagStorageAccountName'))).primaryEndpoints.blob]"
+          }
+        }
+      },
+      "tags": {
+        "displayName": "Web VMs"
+      }
+    },
+    {
+      "type": "Microsoft.Compute/virtualMachines/extensions",
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move to child resource

File: barracuda-waf-iis/azuredeploy.json [hunk 15]
```diff
@@ -0,0 +1,653 @@
...
+        "publisher": "Microsoft.Powershell",
+        "settings": {
+          "ModulesUrl": "[concat(parameters('_artifactsLocation'),'scripts/',variables('dscfilename'),parameters('_artifactsLocationSasToken'))]",
+          "ConfigurationFunction": "[variables('webVmExtensionFunction')]",
+          "wmfVersion": "4.0",
+          "Properties": {
+
+          }
+        },
+        "type": "DSC",
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.json]: move above settings property

File: barracuda-waf-iis/azuredeploy.parameters.json [hunk 16]
```diff
@@ -0,0 +1,30 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "adminUsername": {
+      "value": "bwafadmin"
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.parameters.json]: use GEN-UNIQUE

File: barracuda-waf-iis/azuredeploy.parameters.json [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    "webVmSize": {
+      "value": "Standard_D1"
+    },
+    "bwafVmSize": {
+      "value": "Standard_D1"
+    },
+    "storageAccountType": {
+      "value": "Standard_LRS"
+    },
+    "_artifactsLocation": {
```
Comment (bmoore-msft) [inline barracuda-waf-iis/azuredeploy.parameters.json]: remove params that have default values

**Comment by amalik99** [comment]
Hi Brian, Thanks for the review, updated the template as per recommendations. Please let me know in case of any other consideration.

Amit

**Pull request merged**